server/src/main/java/io/druid/server/coordinator/HttpLoadQueuePeon.java

```diff
@@ -0,0 +1,603 @@
+      }
+    } else {
+      log.debug("[%s] No segments to load/drop.");
+      mainLoopInProgress.set(false);
+    }
+  }
+
+  @Override
+  public void start()
+  {
+    synchronized (lock) {
+      if (stopped) {
+        throw new ISE("Can't start.");
+      }
+
+      ScheduledExecutors.scheduleAtFixedRate(
+          processingExecutor,
+          new Duration(config.getHttpLoadQueuePeonRepeatDelay()),
+          new Callable<ScheduledExecutors.Signal>()
+          {
+            @Override
+            public ScheduledExecutors.Signal call()
+            {
+              if (!stopped) {
+                doSegmentManagement();
+              }
+
+              if (stopped) {
+                return ScheduledExecutors.Signal.STOP;
+              } else {
+                return ScheduledExecutors.Signal.REPEAT;
+              }
+            }
+          }
+      );
+    }
+  }
+
+  @Override
+  public void stop()
+  {
+    synchronized (lock) {
+      if (stopped) {
+        return;
+      }
+
+      stopped = true;
+
+      for (SegmentHolder holder : segmentsToDrop.values()) {
+        holder.requestCompleted(false, null);
+      }
+
+      for (SegmentHolder holder : segmentsToLoad.values()) {
+        holder.requestCompleted(false, null);
+      }
+
+      segmentsToDrop.clear();
+      segmentsToLoad.clear();
+      queuedSize.set(0L);
+      failedAssignCount.set(0);
+    }
+  }
+
+  @Override
+  public void loadSegment(DataSegment segment, LoadPeonCallback callback)
+  {
+    synchronized (lock) {
+      SegmentHolder holder = segmentsToLoad.get(segment);
+
+      if (holder == null) {
+        log.info("Server[%s] to load segment[%s] queued.", serverId, segment.getIdentifier());
+        segmentsToLoad.put(segment, new SegmentHolder(segment, LOAD, callback));
+        processingExecutor.execute(this::doSegmentManagement);
+      } else {
+        holder.addCallback(callback);
+      }
+    }
+  }
+
+  @Override
+  public void dropSegment(DataSegment segment, LoadPeonCallback callback)
+  {
+    synchronized (lock) {
+      SegmentHolder holder = segmentsToDrop.get(segment);
+
+      if (holder == null) {
+        log.info("Server[%s] to drop segment[%s] queued.", serverId, segment.getIdentifier());
+        segmentsToDrop.put(segment, new SegmentHolder(segment, DROP, callback));
+        processingExecutor.execute(this::doSegmentManagement);
+      } else {
+        holder.addCallback(callback);
+      }
+    }
+  }
+
+  @Override
+  public Set<DataSegment> getSegmentsToLoad()
+  {
+    return segmentsToLoad.keySet();
+  }
+
+  @Override
+  public Set<DataSegment> getSegmentsToDrop()
+  {
+    return segmentsToDrop.keySet();
+  }
+
+  @Override
+  public long getLoadQueueSize()
+  {
+    return queuedSize.get();
+  }
+
+  @Override
+  public int getAndResetFailedAssignCount()
+  {
+    return failedAssignCount.getAndSet(0);
+  }
+
+  @Override
+  public void markSegmentToDrop(DataSegment dataSegment)
```

Comment:
What's the difference between this and dropSegment?

Reply:
this behavior is retained from existing behavior in CuratorLoadQueuePeon (copy of older LoadQueuePeon class ). I am not sure about this but it was recently added so kept.